### ISSUE TITLE
DocumentLockerTest / FSSynchronizationTest are flaky

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-eclipse#681]
```
Fehlermeldung

sleep interrupted

Stacktrace

java.lang.InterruptedException: sleep interrupted
	at java.lang.Thread.sleep(Native Method)
	at org.eclipse.xtext.ui.tests.editor.model.DocumentLockerTest.testPriorityReadOnlyCancelsReaders(DocumentLockerTest.java:151)

```

```
java.lang.NullPointerException: null
	at org.eclipse.xtext.builder.impl.AbstractFSSynchronizationTest.testCleanUpDerivedResourcesWithCreateBetween(AbstractFSSynchronizationTest.java:341)
	at org.eclipse.xtext.builder.impl.AbstractFSSynchronizationTest.testCleanUpDerivedResourcesWithCreateBetween(AbstractFSSynchronizationTest.java:255)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
......
```

**Comment by cdietrich** [comment]
the slower the (virtual) hardware the more often this seems to happen.
i propose to ignore the tests for now and investigate what happens

**Comment by ArneDeutsch** [comment]
Here is the stack trace that currently leads to the (sometimes) failing test:

```
java.lang.InterruptedException: null
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1302)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at org.eclipse.xtext.ui.tests.editor.model.DocumentLockerTest.testPriorityReadOnlyCancelsReaders(DocumentLockerTest.java:151)
```

It looks as if the test runner thread is interrupted. I can't say why. But it might be that we are bitten by a JUnit bug here: https://github.com/junit-team/junit4/issues/1365

I will insert a `Thread.currentThread.interrupted` at the start of the test case to ensure the interrupted flag is cleared.

**Comment by ArneDeutsch** [comment]
I started investigating FSSynchronizationTest. To be honest I do not understand what it is testing. The tests are deeply flawed and only succeed BECAUSE they are flawed. The test case mentioned in this ticket `testCleanUpDerivedResourcesWithCreateBetween` calls the method `testCleanUpDerivedResourcesWithCreateBetween(IContainer output)`. Placing a breakpoint at line 281 and running in debug mode makes the test failing all the time. The reason is, that after a clean build the `src-gen` folder contains three files (`.Foo.buildertestlanguagebin`, `Foo.txt` and `Lalala.txt`) wheras the test case expects to find only one (`Lalala.txt`). It seems that some caching or similar mechanism hinders the test case to actually see the new files when it runs at full speed.

I can see following causes wheras I do not know yet what is correct:
# The implementation is broken and the test just does not detect this (in this case the clean build action does not clean the `src-gen` folder as expected and we have to fix it in addition to the flawed test)
# The implementation is correct and the expectation of the test is wrong, but because the test is flawed it succeeds most of the time anyway (in this case we have to fix the test expectations and the check if the files are there or not)

**Comment by cdietrich** [comment]
@szarnekow do you have any hints

**Comment by ArneDeutsch** [comment]
By the way ... the seperation of `AbstractFSSynchronizationTest` and `FSSynchronizationTest` is completly mysterious to me. Someone has an idea what is the reason for the split?

**Comment by cdietrich** [comment]
after the clean build there is another full build which recreates the file

```
xtext builder 6
xtext builder 6
builder done
xtext builder 9
generate file Foo.txt
clean done
builder done
xtext builder 10
builder done
[.Foo.buildertestlanguagebin, Foo.txt]
new touch
[Lalala.txt, .Foo.buildertestlanguagebin, Foo.txt]
builder clean
clean output
clean derived
[L/myProject/src-gen/.Foo.buildertestlanguagebin, L/myProject/src-gen/Foo.txt]
deleteL/myProject/src-gen/.Foo.buildertestlanguagebin
deleteL/myProject/src-gen/Foo.txt
clean done
builder clean done
xtext builder 6
generate file Foo.txt
clean done
builder done
xtext builder 10
builder done
[Lalala.txt, .Foo.buildertestlanguagebin, Foo.txt]
```

**Comment by cdietrich** [comment]
enabled core resources tracing

```
| Worker-3: Building workspace | 2018-05-25 18:17:55.108 | org.eclipse.core.resources | /debug | org.eclipse.core.internal.utils.Policy | debug | 130 | org.eclipse.core.internal.events.AutoBuildJob(Building workspace): Top-level build-start of: myProject; [], INCREMENTAL_BUILD |
| Worker-3: Building workspace | 2018-05-25 18:17:55.109 | org.eclipse.core.resources | /debug | org.eclipse.core.internal.utils.Policy | debug | 130 | org.eclipse.core.internal.events.AutoBuildJob(Building workspace): Invoking (FULL_BUILD) on builder: XtextBuilder(myProject; []) |
``` 

as you can see buildmanager turns incremental build into full build if there is no state
which is always after a clean build i think.

**Comment by cdietrich** [comment]
=> does it make sense to do a clean only build if this will trigger a autobuild anyway?
(see clean dialog with autobuild enabled vs with autobuild disabled)

**Comment by cdietrich** [comment]
=>

with

```
IResourcesSetupUtil.autobuild = false
cleanBuild
``` 
green

but still the question what is the intention of the test

**Comment by ArneDeutsch** [comment]
Althought I think the test code is quite bad (hard to understand, buggy, flaky ...) I would fix it as you suggest. If it then stops beeing flaky I think we have restored the initial indent of the test (guess it should check that the "clean" does remove the generated files in "src-gen" but does NOT remove other files from "src-gen" that the user has created in that folder ... in effect that not simply the complete folder content is dropped).

**Comment by cdietrich** [comment]
@ArneDeutsch looks promising now. i give it a second run on Jenkins to be "more sure"

**Comment by cdietrich** [comment]
Fixed in 2.15
